File: 2016/1/part2.js
const fs = require('fs');

let heading = 0
let location = [0,0]
let dir = [
  [0,1],
  [1,0],
  [0,-1],
  [-1,0]
]
let data = fs.readFileSync('./input', 'utf8')
  .split(',')

console.log(data)
let history = []

process:
for(e of data) {
  let turn = e.match(/[RL]/)[0]
  switch(turn) {
    case 'R':
      heading++
      heading = heading < 4 ? heading : 0
      break
    case 'L':
      heading--
      heading = heading >= 0 ? heading : 3
      break
  }

  let delta = parseInt(e.match(/[0-9]+/)[0])
  let deltac = dir[heading].map(e=>e*delta)

  let increment = dir[heading]
  for(let i = 0; i < delta; i++) {
    Math.abs(deltac[0]) > 0 ? location[0]+= increment[0] : location[1] += increment[1]
    let filter = history.filter(e => e[0] == location[0] && e[1] == location[1])
    if(filter.length>0) {
      console.log(`REVISIT ${location}`)
      console.log(Math.abs(filter[0][0]) + Math.abs(filter[0][1]))
      break process
    }
    
    history.push([location[0],location[1]])
  }
}

File: 2016/1/part2.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int main(void) {
  FILE *input;
  input = fopen("input", "r");

  char str1[8];
  int heading = 0;
  int location[2] = {0,0};
  int dir[4][2] = {
    {0,1},
    {1,0},
    {0,-1},
    {-1,0}
  };
  char card;
  int delta, deltax;
  int (*increment)[2];
  int history[151][2];
  int historyLen = 0;
  
  while(fgets(str1, 8, input)) {
    sscanf(str1, "%c%d,", &card, &delta);
    puts("------------------------------------");
    printf("Turn %c and proceed %d\n", card, delta);
    printf("    location: %d,%d\n",location[0],location[1]);
    switch (card) {
      case 'R':
        puts("    TURNS RIGHT");
        heading++;
        heading = heading < 4 ? heading : 0;
        break;
      case 'L':
        puts("    TURNS LEFT");
        heading--;
        heading = heading >= 0 ? heading : 3;
        break;
    }
    printf("    delta: %d\n", delta);


    deltax = (dir[heading][0] * delta);
    increment = dir[heading];
    for(int i = 0; i < delta; i++) {
        if (abs(deltax) > 0) 
            location[0] += (*increment)[0];
        else
            location[1] += (*increment)[1];

        for (int j = 0; j < historyLen; j++) {
            if (history[j][0] == location[0] && history[j][1] == location[1]) {
                printf("\nREVISIT! %d,%d\n", location[0], location[1]);
                fclose(input);
                exit(0);
            }
        }

        historyLen++;
        history[historyLen][0] = location[0];
        history[historyLen][1] = location[1];
    }
  } 


  printf("FINAL LOCATION: %d,%d\n",location[0],location[1]);
  fclose(input);
}